File: android/sample-ui/src/test/kotlin/com/usesmileid/sampleapps/ui/state/UseSmileIDSampleFlowResultTest.kt
package com.usesmileid.sampleapps.ui.state

import androidx.compose.runtime.saveable.SaverScope
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleEnvironment
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleFlowRoute
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleFlowStatus
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleScenario
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test

class UseSmileIDSampleFlowResultTest {

    @Test
    fun a_fresh_run_has_counted_nothing() {
        val result = UseSmileIDSampleFlowResult().snapshot
        assertEquals(0, result.resultCallbackCount)
        assertEquals(0, result.refreshCallbackCount)
        assertEquals(UseSmileIDSampleFlowStatus.Idle, result.jobStatus)
    }

    @Test
    fun each_callback_counts_once() {
        val flow = UseSmileIDSampleFlowResult()
        flow.recordRefreshCallback()
        flow.recordRefreshCallback()
        flow.recordResultCallback(UseSmileIDSampleFlowStatus.Succeeded, jobId = "job_1", userId = "user_1")

        assertEquals(1, flow.resultCallbackCount)
        assertEquals(2, flow.refreshCallbackCount)
        assertEquals("job_1", flow.jobId)
        assertEquals("user_1", flow.userId)
    }

    @Test
    fun a_second_result_callback_is_visible_rather_than_swallowed() {
        val flow = UseSmileIDSampleFlowResult()
        flow.recordResultCallback(UseSmileIDSampleFlowStatus.Succeeded, jobId = "job_1")
        flow.recordResultCallback(UseSmileIDSampleFlowStatus.Succeeded, jobId = "job_1")
        assertEquals(2, flow.resultCallbackCount)
    }

    @Test
    fun starting_a_run_clears_the_previous_outcome_and_both_counts() {
        val flow = UseSmileIDSampleFlowResult()
        flow.recordRefreshCallback()
        flow.recordResultCallback(UseSmileIDSampleFlowStatus.Failed, error = "2213: authentication failed")

        flow.startFlow(UseSmileIDSampleEnvironment.Production)

        assertEquals(UseSmileIDSampleFlowStatus.Running, flow.status)
        assertEquals("the run's own environment, from its snapshot", UseSmileIDSampleEnvironment.Production, flow.environment)
        assertEquals(0, flow.resultCallbackCount)
        assertEquals(0, flow.refreshCallbackCount)
        assertNull(flow.lastError)
        assertNull(flow.jobId)
    }

    @Test
    fun recreation_preserves_the_counts_and_the_outcome() {
        val flow = UseSmileIDSampleFlowResult(scenario = UseSmileIDSampleScenario.BadRefresh)
        flow.enterRoute(UseSmileIDSampleFlowRoute.Shell)
        flow.recordRefreshCallback()
        flow.recordResultCallback(UseSmileIDSampleFlowStatus.Failed, error = "2213: authentication failed")

        assertEquals(flow.snapshot, flow.recreate().snapshot)
    }

    @Test
    fun a_scenario_this_build_no_longer_has_restores_as_the_default() {
        val restored = UseSmileIDSampleFlowResult.Saver.restore(
            listOf("retiredScenario", "brandDefault", "fullscreen", "sandbox", "idle", "", "", "", "0", "0"),
        )
        assertEquals(UseSmileIDSampleScenario.Normal, restored?.scenario)
    }

    @Test
    fun an_unreadable_environment_restores_as_sandbox_rather_than_taking_the_app_down() {
        val restored = UseSmileIDSampleFlowResult.Saver.restore(
            listOf("normal", "brandDefault", "fullscreen", "somewhere-else", "idle", "", "", "", "0", "0"),
        )
        assertEquals(UseSmileIDSampleEnvironment.Sandbox, restored?.environment)
    }

    @Test
    fun a_count_that_did_not_round_trip_restores_as_zero() {
        val restored = UseSmileIDSampleFlowResult.Saver.restore(
            listOf("normal", "brandDefault", "fullscreen", "production", "idle", "", "", "", "", "not-a-number"),
        )
        assertEquals(0, restored?.resultCallbackCount)
        assertEquals(0, restored?.refreshCallbackCount)
    }

    private fun UseSmileIDSampleFlowResult.recreate(): UseSmileIDSampleFlowResult {
        val saved = with(UseSmileIDSampleFlowResult.Saver) { SaverScope { true }.save(this@recreate) }
        assertNotNull("the saver produced nothing to restore from", saved)
        return requireNotNull(UseSmileIDSampleFlowResult.Saver.restore(saved!!))
    }
}

File: ios/SampleUI/Tests/SampleUITests/UseSmileIDSampleFlowResultTest.swift
@testable import SampleUI
import XCTest

final class UseSmileIDSampleFlowResultTest: XCTestCase {
  func testAFreshRunHasCountedNothing() {
    let result = UseSmileIDSampleFlowResult().snapshot
    XCTAssertEqual(result.resultCallbackCount, 0)
    XCTAssertEqual(result.refreshCallbackCount, 0)
    XCTAssertEqual(result.jobStatus, .idle)
  }

  func testEachCallbackCountsOnce() {
    var flow = UseSmileIDSampleFlowResult()
    flow.recordRefreshCallback()
    flow.recordRefreshCallback()
    flow.recordResultCallback(status: .succeeded, jobId: "job_1", userId: "user_1")

    XCTAssertEqual(flow.resultCallbackCount, 1)
    XCTAssertEqual(flow.refreshCallbackCount, 2)
    XCTAssertEqual(flow.jobId, "job_1")
    XCTAssertEqual(flow.userId, "user_1")
  }

  func testASecondResultCallbackIsVisibleRatherThanSwallowed() {
    var flow = UseSmileIDSampleFlowResult()
    flow.recordResultCallback(status: .succeeded, jobId: "job_1")
    flow.recordResultCallback(status: .succeeded, jobId: "job_1")
    XCTAssertEqual(flow.resultCallbackCount, 2)
  }

  func testStartingARunClearsThePreviousOutcomeAndBothCounts() {
    var flow = UseSmileIDSampleFlowResult()
    flow.recordRefreshCallback()
    flow.recordResultCallback(status: .failed, error: "2213: authentication failed")

    flow.startFlow(environment: .production)

    XCTAssertEqual(flow.status, .running)
    XCTAssertEqual(flow.environment, .production, "the run's own environment, from its snapshot")
    XCTAssertEqual(flow.resultCallbackCount, 0)
    XCTAssertEqual(flow.refreshCallbackCount, 0)
    XCTAssertNil(flow.lastError)
    XCTAssertNil(flow.jobId)
  }

  func testRecreationPreservesTheCountsAndTheOutcome() {
    var flow = UseSmileIDSampleFlowResult(scenario: .badRefresh)
    flow.enterRoute(.shell)
    flow.recordRefreshCallback()
    flow.recordResultCallback(status: .failed, error: "2213: authentication failed")

    XCTAssertEqual(UseSmileIDSampleFlowResult(saved: flow.saved).snapshot, flow.snapshot)
  }

  func testAScenarioThisBuildNoLongerHasRestoresAsTheDefault() {
    let restored = UseSmileIDSampleFlowResult(
      saved: ["retiredScenario", "brandDefault", "fullscreen", "sandbox", "idle", "", "", "", "0", "0"]
    )
    XCTAssertEqual(restored.scenario, .normal)
  }

  func testAnUnreadableEnvironmentRestoresAsSandboxRatherThanTakingTheAppDown() {
    let restored = UseSmileIDSampleFlowResult(
      saved: ["normal", "brandDefault", "fullscreen", "somewhere-else", "idle", "", "", "", "0", "0"]
    )
    XCTAssertEqual(restored.environment, .sandbox)
  }

  func testACountThatDidNotRoundTripRestoresAsZero() {
    let restored = UseSmileIDSampleFlowResult(
      saved: ["normal", "brandDefault", "fullscreen", "production", "idle", "", "", "", "", "not-a-number"]
    )
    XCTAssertEqual(restored.resultCallbackCount, 0)
    XCTAssertEqual(restored.refreshCallbackCount, 0)
  }

  /// A blocked run is a failure the SDK never saw, so the result count must not move.
  func testABlockedRunFailsWithoutCountingAResultCallback() {
    var flow = UseSmileIDSampleFlowResult()
    flow.recordBlocked(reason: "Scan a token first", environment: .production)
    XCTAssertEqual(flow.status, .failed)
    XCTAssertEqual(flow.lastError, "Scan a token first")
    XCTAssertEqual(flow.environment, .production)
    XCTAssertEqual(flow.resultCallbackCount, 0)
  }
}
